File: two_pointers/239.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var maxSlidingWindow = function (nums, k) {
    const n = nums.length
    const deq = []
    const result = []
    for (let i = 0; i < n; i++) {
        while (deq.length && deq[0] <= i - k) {
            deq.shift()
        }
        while (deq.length && nums[i] > nums[deq[deq.length - 1]]) {
            deq.pop()
        }
        deq.push(i)
        if (i >= k - 1) {
            result.push(nums[deq[0]])
        }
    }
    return result
};

File: two_pointers/239.py


# Tc = O(n) [only visiting twice each element, only push and pop]

from typing import List
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        n = len(nums)
        deq = deque()     # will store indexes (not values)
        result = []
        
        for i in range(n):
            while deq and deq[0] <= i - k:
                deq.popleft()  # remove elements out of this window
                
            while deq and nums[i] > nums[deq[-1]]:
                deq.pop()  # remove smaller numbers as they can't be max
                
            deq.append(i)  # add current index
            
            if i >= k - 1:
                result.append(nums[deq[0]])  # first index in deque is max
        
        return result


            